### ISSUE TITLE
There is displacement of pixel between input and output

### ISSUE THREAD

**Issue by cfanyyx** [chaofengc/PSFRGAN#21]
Excellent work! But I find that there is some displacement of pixel between input and output. Can this be fixed?

**Comment by chaofengc** [comment]
Would you please show some examples ?

**Comment by cfanyyx** [comment]
For example, input and output below, if you pageUp and pageDown those two figures, you will find the portrait moves a little bit.
![001](https://user-images.githubusercontent.com/7723243/110059483-47095c80-7d9f-11eb-90d7-1f4e21acfb82.jpg)
![000](https://user-images.githubusercontent.com/7723243/110059495-4bce1080-7d9f-11eb-980c-b045c6933e00.jpg)

**Comment by chaofengc** [comment]
Thanks for pointing out this problem. However, there are no easy solution to it.
The misalignment is likely because the synthesized LQ images are not perfectly aligned with HQ images after the degradation and upsample process (possibly because of motion blur). And the network somewhat learns to shift the inputs a little bit.

**Comment by cfanyyx** [comment]
But I find DFDNet and HiFaceGAN don't have this problem, which use the similar degradation strategy.

**Comment by chaofengc** [comment]
This might because of different hyper parameters. I will try to train a model without motion blur.

**Comment by chaofengc** [comment]
Thanks again for helping us to find this problem. After carefully debugging, we found that the reason is that the nearest resize of pytorch, i.e. `nn.functional.interpolate`, does not really do nearest interpolate as opencv and PIL. You may refer to this issue for details: https://github.com/pytorch/pytorch/issues/34808#issuecomment-638469752. Because it only pick the pixel in the left top corner, the results are moved to the left top direction a little bit. DFDNet and HiFaceGAN both have skip connections, and use `max_pool` and `avg_pool` to do the downsample, therefore do not have this problem.

The reason we use nearest downsample is because we do not want to introduce new labels to parsing map. And we did not notice the implementation problem of nearest downsample in pytorch. 

To verify the above conclusion, you may replace this line in the code with the following reimplemented nearest downsample function (**need to calculate scale by yourself**):
https://github.com/chaofengc/PSFRGAN/blob/5fc0e1f616b5b87ebdd96f450c53b60d4a3c5c4b/models/psfrnet.py#L32
```
def nearest_resize_pil_impl(org, scale=0.5):
    b, c, h, w = org.shape
    resized = torch.zeros((b, c, h//2, w//2)).to(org)
    for i in range(resized.shape[2]):
        for j in range(resized.shape[3]):
            x = scale * (j + 0.5) - 0.5 
            y = scale * (i + 0.5) - 0.5
            ii = int(y)
            jj = int(x)
            resized[:, :, i, j] = org[:, :, ii, jj]
    return resized
```

And you may observe that the misalignment is gone, but the results are corrupted. We will retrain our model, and released the improved version asap.

**Comment by chaofengc** [comment]
Thanks again for helping us finding this problem. After carefully debugging, we found training data is not the problem. We suspect there are mainly two reasons:

- The nearest resize of pytorch, i.e. `nn.functional.interpolate`, does not really do nearest interpolate as opencv and PIL. You may refer to this issue for details: https://github.com/pytorch/pytorch/issues/34808#issuecomment-638469752. Because it only pick the pixel in the left top corner, the results are moved to the left top direction a little bit. DFDNet and HiFaceGAN both have skip connections, and use `max_pool` and `avg_pool` to do the downsample, therefore do not have this problem.
https://github.com/chaofengc/PSFRGAN/blob/5fc0e1f616b5b87ebdd96f450c53b60d4a3c5c4b/models/psfrnet.py#L32

- The predicted parsing map from LQ inputs does not exactly correct in pixel level.

We will try to solve the first problem and release the improved model asap. For the second problem, there is no easy solution to predict accurate parsing map for LQ faces. We currently have no plan to improve the pretrained face parsing network.

**Comment by chaofengc** [comment]
We finally found that using `bicubic` interpolation to replace `nearest` interpolation gives the best results. Here is the preliminary result for your image:
![test_align](https://user-images.githubusercontent.com/7187685/110726509-795e0280-8254-11eb-91aa-a519bc139519.jpg)

I will release the final model when the training finished.

**Comment by cfanyyx** [comment]
Great! Hope to see the new model soon.

**Comment by cfanyyx** [comment]
Another question, I want to know whether you get the new model by finetuning the old one or training from begining with a set of new training params(or new dataset)? Because I find the style of new result is a little different from the old one(for example, the eyes).

**Comment by chaofengc** [comment]
The new model is trained from scratch with the same dataset. Some small changes are made to the codes as well as the degradation parameters.  Because the results of GAN is unstable, model at different checkpoints might give results with different styles. 
I will release the new model together with the training codes. You may try to train your own model.

**Comment by cfanyyx** [comment]
OK, thanks.

**Comment by cfanyyx** [comment]
@chaofengc Sorry to bother you again. After reproducing your work. I want to ask a few other questions.
1.How did you fix the nose problem(chromatic aberration) in your new model? And how did you fix the chromatic aberration between the output of network and original pic? Did you use some Matching loss to correct this problem?
2.I find it is very likely to see the artifacts in the hair part of my results. Have you encountered this problem and how did you fix it?

**Comment by chaofengc** [comment]
We did not have this problem during training. We will release the training codes soon, please stay tuned.

**Comment by cfanyyx** [comment]
The old model indeed has the chromatic aberration problem.
![image](https://user-images.githubusercontent.com/7723243/111111691-c7973c80-8599-11eb-8518-00533300833c.png)
![image](https://user-images.githubusercontent.com/7723243/111112592-5d7f9700-859b-11eb-8096-f40b0b8a6d34.png)

**Comment by chaofengc** [comment]
The model is trained to mainly deal with RGB photos. We use `RGBtoGray` to synthesis old gray photos but this does not apply to all old photos.

**Comment by cfanyyx** [comment]
But this problem seems alleviated in the new model from the result you posted.

**Comment by chaofengc** [comment]
The difference is mostly because of the unstable training of GAN. The new model is a selected better model. The results are slightly better than the old one but also bad in some other cases. 

You will have to improve the training data if you want the model to work for most real cases.

**Comment by cfanyyx** [comment]
Got it!
